index.js

```diff
@@ -15,51 +15,198 @@
+var aproba = require('aproba');
+var bcp47 = require('bcp47');
+var bcp47stringify = require('bcp47-stringify');
 
-'use strict';
-var engine = require('adaro'),
-    munger = require('./lib/munger');
+/**
+ * Make a View class that uses our configuration, set far in advance of
+ * instantiation because Express passes very little to the actual constructor.
+ */
+function makeViewClass(config) {
+    aproba('O', arguments);
 
+    var conf = normalizeConfigs(config);
 
-exports.dust = function (setting, config) {
-    var settings =  (arguments.length > 1) ? setting : {},
-        configs = (arguments.length > 1) ? config : setting,
-        renderer;
+    var proto = Object.create(oldView.prototype);
 
-    if (!configs || !(configs.specialization || configs.i18n)) {
-        return engine.dust(settings);
-    }
+    // Unfortunately, since we don't know the actual file to resolve until
+    // we get request context (in `render`), we can't say whether it exists or not.
+    //
+    // Express checks that this is truthy to see if it should return an error or
+    // run the render, so we hard code it to true.
+    proto.path = true;
+
+    proto.lookup = function lookup(name, options, cb) {
+        if (arguments.length === 1) {
+            // This is the unoverriden constructor calling us. Ignore the call.
+            return true;
+        }
+
+        var ext = path.extname(name);
+
+        if (conf[ext] && conf[ext].specialization) {
+            var nameNoExt = name.slice(0, -ext.length);
+            var newName = conf[ext].specialization.resolve(nameNoExt, options) + ext;
+            debug("specialization mapped '%s' to '%s'", name, newName);
+            name = newName;
+        }
+
+        var search = [];
+        search.push([].concat(conf[ext] && conf[ext].root ? conf[ext].root : this.root));
+
+        if (conf[ext] && conf[ext].i18n) {
+            var i18n = conf[ext].i18n;
+            var locales = [];
+            if (options.locale) {
+                locales.push(i18n.formatPath(typeof options.locale === 'object' ? options.locale : bcp47.parse(options.locale.replace(/_/g, '-'))));
+            }
+            if (i18n.fallback) {
+                locales.push(i18n.formatPath(i18n.fallback));
+            }
+            debug("trying locales %j", locales);
+            search.push(locales);
+        }
+
+        search.push([name, path.join(path.basename(name), 'index' + ext)]);
+
+        debug('lookup "%s"', name);
+
+        permutron.raw(search, function (candidate, next) {
+            var resolved = path.resolve.apply(null, candidate);
+            limitStat(resolved, function (err, stat) {
+                if (!err && stat.isFile()) {
+                    debug('found "%s"', resolved);
+                    cb(null, resolved);
+                } else if ((!err && stat.isDirectory()) || (err && err.code === 'ENOENT')) {
+                    next();
+                } else {
+                    cb(err);
```

Comment:
Yeah. I'm not super happy with that logic -- it works, it's correct, but it's not obvious.